x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ExtractedField.java

```diff
@@ -48,10 +50,12 @@ enum Method {
 
     /**
      * Extracts the value from a {@link SearchHit}
-     * @param hit the search hit
+     *
+     * @param hit    the search hit
+     * @param source the source supplier with caching capabilities
      * @return the extracted value
      */
-    Object[] value(SearchHit hit);
+    Object[] value(SearchHit hit, Supplier<Map<String, Object>> source);
```

Comment:
is it up to the caller in these cases to provide a supplier that may or may not cache, or is the expectation that caching should always happen (or the other way around)?

Reply:
The caller should provide the supplier and it may never be used. It is only used if the extracted field is a source field.

Reply:
And is the supplier always doing the caching when provided?

Reply:
No, only if it  used: https://github.com/elastic/elasticsearch/blob/12c23bc6db9ca2b113a98fec67f9e1aede39ce6c/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/SourceSupplier.java

Reply:
right, then I wonder if it needs to be as generic as a supplier, if there is a single impl ever used. it could be something more specific to caching in that that supplier provides caching.

Reply:
I agree, I will change the API to get a more specific SourceSupplier

Reply:
thanks!